src/main/java/hudson/remoting/Engine.java

```diff
@@ -203,14 +206,27 @@ public Thread newThread(final Runnable r) {
 
     private DelegatingX509ExtendedTrustManager agentTrustManager = new DelegatingX509ExtendedTrustManager(new BlindTrustX509ExtendedTrustManager());
 
+    private String directUrl;
+    private String instanceIdentity;
+    private final Set<String> protocols;
+
     public Engine(EngineListener listener, List<URL> hudsonUrls, String secretKey, String slaveName) {
+        this(listener, hudsonUrls, secretKey, slaveName, null, null, null);
+    }
+
+    public Engine(EngineListener listener, List<URL> hudsonUrls, String secretKey, String slaveName, String directUrl, String instanceIdentity,
+                  Set<String> protocols) {
         this.listener = listener;
+        this.directUrl = directUrl;
         this.events.add(listener);
         this.candidateUrls = hudsonUrls;
         this.secretKey = secretKey;
         this.slaveName = slaveName;
-        if(candidateUrls.isEmpty())
+        this.instanceIdentity = instanceIdentity;
+        this.protocols = protocols;
+        if(candidateUrls.isEmpty() && instanceIdentity == null) {
```

Comment:
Shouldn't we also check `directUrl` and `instanceIdentity` to fail early if not set properly? Something like:
```
if(isEmpty(directUrl) ^ isEmpty(instanceIdentity)) {
    throw new IllegalArgumentException(
        "Both 'direct' and 'instanceIdentity' need to be specified for direct connections");
} else if(candidateUrls.isEmpty() && isEmpty(directUrl)) {
...
```

Reply:
Take a look at how I defined the "@Options", here and in nearby lines: https://github.com/jenkinsci/remoting/pull/338/files#diff-60b1d0697cd30e065bf0996f3b1ff438R183

The "depends" and "forbids" values should handle that in a clean way. We might could tighten that up a, especially with existing arguments but I think I got it pretty good for the new ones.

I tested the behavior at one point and it did pretty much what I expected. Probably worth testing again that I got it lined up correctly.

Reply:
I saw the "depends" and "forbids" in `Main` &ndash; nice.
But if remoting is used as library `Engine` might be used directly, bypassing `Main`.

Reply:
True, but I don't know if it's worth duplicating / complicating the code for what is probably a rare situation. This code is already messy enough. It really needs to be reworked with better structured class responsibility and isolation, but that introduces API changes. If people think it's important to duplicate the check I can add it.